avoid code duplication between this RunMojo and StartMojo

simonjenga opened the pull request:
#42 avoid code duplication between this RunMojo and StartMojo
@dmarkov Please assign someone this PR and let me know if this is what was expected.

dmarkov commented:
@simonjenga Let me find a reviewer for this pull request, thanks for submitting it

dmarkov commented:
@dmzaytsev it's yours, please review

dmzaytsev commented:
@simonjenga Travis build failed
let's fix it

simonjenga commented:
@dmzaytsev I have fixed the Travis build failure ```5 PMD violations```.
Please review.

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractDynamoMojo.java:
```diff
@@ -96,6 +96,8 @@ public void execute() throws MojoFailureException {
             Logger.info(this, "execution skipped because of 'skip' option");
             return;
         }
+        // @checkstyle Always make this.environment(); as the first call!
```
@simonjenga why do you use `@checkstyle` annotation here?
if you really need this comment, put it into javadoc

simonjenga replied:
Ok

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractDynamoMojo.java:
```diff
@@ -120,11 +122,20 @@ protected int tcpPort() {
+     * {@link com.jcabi.dynamodb.maven.plugin.AbstractEnviromentMojo}.
+     * @throws MojoFailureException If fails
+     */
+    protected void environment() throws MojoFailureException {
```
@simonjenga we can use abstract method, right? like `#run()`

simonjenga replied:
If you make ```#environment``` to be an abstract method, then other sub classes ```com.jcabi.dynamodb.maven.plugin.StopMojo``` and ```com.jcabi.dynamodb.maven.plugin.CreateTablesMojo``` MUST implement it.
This is not what we want because these sub classes do not need to set up an environment for them. Only the sub classes that need environment implement it.

dmzaytsev replied:
@simonjenga ok, let's leave it as it is

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractEnviromentMojo.java:
```diff
@@ -0,0 +1,99 @@
+            );
+        }
+        if (this.home == null) {
+            this.home = new File(System.getProperty("java.home"));
```
@simonjenga what happen  if property `java.home` din't exists

simonjenga replied:
This is how the original Author of that property intended it to be.
```
@Parameter(required = false)
private transient File home;
```

dmzaytsev replied:
@simonjenga I mean if system property `java.home` doesn't set 
in this case `System.getProperty("java.home")`will return null
so `new File (null)` will raise an exception, isn't it?

simonjenga replied:
@dmzaytsev I have caught that exception.

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/RunMojo.java:
```diff
@@ -29,77 +29,41 @@
+ * @author Simon Njenga (simtuje@gmail.com)
  * @version $Id$
  * @since 0.8
+ * @todo #41:30min Let's avoid code duplication between this class and `StartMojo`.
```
@simonjenga the puzzle must  be removed
btw why you didn't create a decorator as the puzzle creator suggested?

simonjenga replied:
```One idea is to create a decorator.``` This is what the creator of the puzzle suggested. But there may be many other ideas. Should we only limit ourselves to only one idea? Let me know in order to make the required changes.

dmzaytsev replied:
@simonjenga we shouldn't
feel free to implement your ideas

dmzaytsev commented:
@simonjenga please see a few comments above

simonjenga commented:
@dmzaytsev See the addressed code review changes. Also, i have included a ```Puzzle``` in ```com.jcabi.dynamodb.maven.plugin.Instances```.

Currently those two methods are similar and only the Architect really knows how they will differ.

I also made changes to make sure the ```run()``` method executes on a daemon/separate thread instead of the current thread.

This task was just to remove duplication.

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractDynamoMojo.java:
```diff
@@ -120,11 +121,20 @@ protected int tcpPort() {
+     * Set the project environment.
+     * {@link com.jcabi.dynamodb.maven.plugin.AbstractEnviromentMojo}.
+     * @throws MojoFailureException If fails
+     * @checkstyle Intentionally empty!, To be implemented by sub classes!
```
@simonjenga why you need `@checkstyle` annotation here ?
please remove the annotation and move the comment up to the second line

simonjenga replied:
@dmzaytsev That annotation is needed to supress the empty method ```return;```

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractEnviromentMojo.java:
```diff
@@ -0,0 +1,106 @@
+        if (this.home == null) {
+            try {
+                this.home = new File(System.getProperty("java.home"));
+             // @checkstyle IllegalCatch (1 line)
```
@simonjenga let's get rid the suppression

simonjenga replied:
@dmzaytsev That supression is needed since we are catching a generic exception.

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractEnviromentMojo.java:
```diff
@@ -0,0 +1,106 @@
+            try {
+                this.home = new File(System.getProperty("java.home"));
+             // @checkstyle IllegalCatch (1 line)
+            } catch (final Exception ex) {
```
@simonjenga File(String) raises NullPointerException
see https://docs.oracle.com/javase/7/docs/api/java/io/File.html#File(java.lang.String)
let's catch NPE instead of Exception

simonjenga replied:
@dmzaytsev It's against java conventions to catch NPE.

dmzaytsev replied:
@simonjenga we shouldn't catch a generic exception.
you catch generic exception and report "Java home property not set", I think it is a wrong way
if you don't want catch NPE you can remove `try` block and check `System.getProperty("java.home")` before calling `File()`

simonjenga replied:
i have removed the ```try`` block.

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/Instances.java:
```diff
@@ -51,6 +51,9 @@
  * @since 0.1
  * @checkstyle ClassDataAbstractionCoupling (500 lines)
  * @see <a href="http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.html">DynamoDB Local</a>
+ * @todo #44:30min Let's avoid code duplication between
```
@simonjenga The issue 44 doesn't exists
I guess you meant 42

simonjenga replied:
@dmzaytsev It's 44 because it's the next issue to be created by the puzzle. The issue 42 already exists!

dmzaytsev replied:
@simonjenga use in the puzzle same issue number.
a new issue will be created automatically with another id
the issue number in `@todo` needs for a trace purpose

simonjenga replied:
@dmzaytsev Ok

dmzaytsev commented:
@simonjenga please see 4 comments above

dmzaytsev commented:
@simonjenga also please fix Travis's build. It failed

simonjenga commented:
@dmzaytsev Please see the feedback review.

simonjenga commented:
@dmzaytsev Please see the feedback review.

simonjenga commented:
@dmzaytsev Please see the feedback review.

dmzaytsev commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractEnviromentMojo.java:
```diff
@@ -0,0 +1,101 @@
+        }
+        // @checkstyle MultipleStringLiterals (2 lines)
+        if (this.home == null) {
+            if (new File(System.getProperty("java.home")).exists()) {
```
@simonjenga my bad 
JRE sets property `java.home`, so this check is redundant
please remove it
in this case we don't need checkstyle suppression two lines above

simonjenga replied:
@dmzaytsev Absolutely this check was redundant! I have reverted back the code to how the Architect has written.

dmzaytsev commented:
@simonjenga just one comment and we are ready to merge

simonjenga commented:
@dmzaytsev Please see the feedback fix.

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev Thanks for your request. @yegor256 Please confirm this.

yegor256 commented on src/main/java/com/jcabi/dynamodb/maven/plugin/AbstractDynamoMojo.java:
```diff
@@ -120,11 +121,20 @@ protected int tcpPort() {
+     * {@link com.jcabi.dynamodb.maven.plugin.AbstractEnviromentMojo}.
+     * @throws MojoFailureException If fails
+     */
+    protected void environment() throws MojoFailureException {
```
@simonjenga why can't you make it `abstract`?

simonjenga replied:
@yegor256 https://github.com/jcabi/jcabi-dynamodb-maven-plugin/pull/43#discussion_r48621997

yegor256 commented:
@simonjenga one comment above

simonjenga commented:
@yegor256 See reply.

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6125-169007740)

rultor commented:
> @rultor merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/6125-169007740) (took me 5min)

dmarkov commented:
@dmzaytsev **35 mins** was added to your account, many thanks for your contribution (payment 73991401)! The task took 167 hours and 53 mins.. extra minutes for review comments (c=20). +35 added to your rating, current score is: [+1369](http://www.netbout.com/b/35166?open=rating)

dmarkov commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@dmarkov OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6125-169632187)

rultor commented:
> @rultor deploy now pls

@dmarkov Done! FYI, the full log is [here](http://www.rultor.com/t/6125-169632187) (took me 5min)
